### ISSUE TITLE
Recover from failed publish

### ISSUE THREAD

**Issue by heruan** [lerna/lerna#455]
I lost connectivity during `lerna publish` and some packages got `lerna-temp` published and other not. Now I have a dirty tree were `lerna publish` fails with:

```
$ lerna publish
Lerna v2.0.0-beta.30
Independent Versioning Mode
Checking for updated packages...
Errored while running PublishCommand.initialize
Error: No updated packages to publish.
    at PublishCommand.initialize (/usr/local/lib/node_modules/lerna/lib/commands/PublishCommand.js:94:18)
    at PublishCommand._attempt (/usr/local/lib/node_modules/lerna/lib/Command.js:161:21)
    at PublishCommand.runCommand (/usr/local/lib/node_modules/lerna/lib/Command.js:145:12)
    at PublishCommand.run (/usr/local/lib/node_modules/lerna/lib/Command.js:69:12)
    at Object.<anonymous> (/usr/local/lib/node_modules/lerna/bin/lerna.js:58:11)
    at Module._compile (module.js:409:26)
    at Object.Module._extensions..js (module.js:416:10)
    at Module.load (module.js:343:32)
    at Function.Module._load (module.js:300:12)
    at Function.Module.runMain (module.js:441:10)
```

How can I recover from this?

**Comment by gigabo** [comment]
@heruan Yikes!  Sorry you ran into trouble.

Lerna uses git tags to keep track of published versions.  It looks like the _tag_ for the version you were in the middle of publishing may have been created before the command failed.

You can see the latest tag with this command:

```
$ git describe --abbrev=0 --tags
v2.0.0-beta.31
```

And you can see the associated commit using the tag name:

```
$ git show v2.0.0-beta.31
```

The tag should refer to a commit on your current branch.  If you haven't made any commits since the failed publish then it should be `HEAD`.

A few things to check:

1. Were any of the packages actually _published_?  You can check with `npm dist-tag ls <package-name>`.  Packages may have been published but may not yet have received the `latest` tag.
1. Were your branch and tag pushed to your remote repo?  The commits may exist only locally.

It's important to know the state of your packages in the npm registry and your repo in your remote repository before proceeding.

**Comment by sccolbert** [comment]
I just got bit by a similar issue. `lerna publish` failed because one of my scoped packages was not yet published with `--access=public` so npm rejected the publish. After correcting that, `lerna` thinks there is nothing updated to publish, but none of the npm tags have yet been updated.

**Comment by KELiON** [comment]
I had the same issue: I've added new package, but when I wanted to publish it, `npm publish` failed with "You should confirm your email before publishing new package" and after that I can't publish new version. 

I had version 0.0.13, started update to 0.0.14 with lerna, but after this fail I see `vundefined` tag was created, but no `v0.0.14`. I tried to create it manually, `git pull --tags`, but after running `lerna publish` I see `vundefined` again. Any thoughts how to fix this "wrong" state?

**Comment by pixelass** [comment]
I always get this issue when I forget to `npm login` before publishing new public packages

**Comment by timneedham** [comment]
@pixelass  and @KELiON... that specific problem went away when I `lerna init` again to get a missing version property in `lerna.json`:

```
{
  "lerna": "2.9.0",
  "packages": [
    "packages/*"
  ],
  "hoist": true,
  "command": {
    "init": {
      "exact": false
    }
  },
  "version": "0.0.1"
}
```

**Comment by rjdestigter** [comment]
I use this script to double check if the packages that are managed by lerna have the latest dependencies as well as if they conflict with npm after publishing fails:

https://gist.github.com/rjdestigter/e673ccfcfff921d40d26246bfbb4ca60

**Comment by vittalas** [comment]
I ran into this issue as well where my lerna publish failed due to some issue with our repository but the tags were created and now i am not able to republish that since it says no updated packages to publish

lerna info Checking for prereleased packages...
lerna info No updated packages to publish. 

what is the solution for this? how can i republish already created tags by lerna

**Comment by gwely** [comment]
When publishing to the registry fails, is it possible to not only not push the "Publish" commit, but to revert the commit and delete the tags? Or just not take these steps until the packages have been successfully published?

**Comment by amphro** [comment]
That would be my vote too. I don't think the "Publish" commit or tag should be pushed if the NPMs don't publish.

**Comment by flodev** [comment]
I had a lerna publish error because of a wrong configured .npmrc
In my case I had to (fix my npmrc first and then) go through all my packages and and execute manually `npm publish --ignore-scripts <packagename.tgz>` (tgz files where already created by lerna publish)

**Comment by tehpsalmist** [comment]
What's the status with this issue? I just had a situation where I tried to `lerna publish` and had some things out of line, so the publishing failed (I don't think it even got to the npm publish part, tbh), but when I check github, I see that some bogus commit was already pushed to github's master branch with the message "Publish." Surely this isn't how it's supposed to work, right? What could I have been doing wrong?

**Comment by absassi** [comment]
I don't know the Lerna's code base, but it seems so trivial of a change to just delay the `git push` to the last step. It's quite simple to revert all other Git changes manually (deleting tags and resetting to previous commit), but once tags are pushed, it can be really hard to delete them (someone else may have already fetched them and they have to manually delete them as well). 

Anyway, at least there's the *hidden* `--no-push` option (it's not in the publish help, but it works), so I'm using `lerna publish --no-push && git push --tags`.

I've seen in other issues suggestions to use `from-package` to just retry without having to undo anything on Git, but, if I understand well, that only works if the issue was temporary, not if there was a bug in your code preventing the publishing to work (e.g. a mistake in some `package.json` that only affects the publish command), because that requires a new commit to fix it, and therefore you normally don't want leave the wrong tags the repository.

**Comment by tehpsalmist** [comment]
Thanks for pointing out `--no-push`, that is useful.

**Comment by kadishmal** [comment]
The way I fixed this issue is as follows:

1. Uncommit from the local repo and tag.

       git reset --hard HEAD^
       git tag -d v1.0.0

2. Remove the last commit and tag in the remote (Github) pushed by Lerna.

       git push origin +HEAD
       git push --delete origin v1.0.0

3. Retry `lerna publish`

       lerna publish

**Comment by JamesHenry** [comment]
Hi Folks 👋

As you may have seen in our published roadmap for Lerna v7 here: https://github.com/lerna/lerna/discussions/3410 recoverable publish is a priority for us in this current phase of work.

I am going to close any duplicate issues in this area and point them here. As a lerna user myself I'm excited to land this!

Many thanks 🙏

**Comment by JamesHenry** [comment]
Hi again, delighted to say that this capability is already available in the latest lerna (`6.5.1` at the time of writing) thanks to #3513 

Please upgrade to the latest and you should not run into republishing issues again.

Many thanks!

**Comment by jonsoku2** [comment]
```shell
#!/bin/sh

# Check if the system is macOS
if [ "$(uname)" != "Darwin" ]; then
  echo "This script is only compatible with macOS."
  exit 1
fi

# Check if jq is installed and available in the PATH
if ! command -v jq >/dev/null 2>&1; then
  echo "jq is not installed or not in the PATH. Please install jq first."
  exit 1
fi

# Define variables
JSON_FILE="./lerna.json"
KEY="version"

# Extract the value of the specified key from the JSON file and save it as a variable
VALUE=v$(cat "$JSON_FILE" | jq -r ".$KEY")

# Print the value of the variable
echo "$VALUE"

# Remove the last commit and tag in the local repo
git reset --hard HEAD^
git tag -d "$VALUE"

# Remove the last commit and tag in the remote (Github) repository
git push origin +HEAD
git push --delete origin "$VALUE"

# Retry lerna publish
npm run publish:libs
```

i made this shell script and using now
